### DISCUSSION TITLE
Regarding the use of vertical alignment, there is a discrepancy between the vertical centering effect of the text and the expected outcome.

### DISCUSSION THREAD

**Discussion by inernoro** [SixLabors/ImageSharp.Drawing#303] [Q&A] [answered]
1. Is there an issue with my usage? My understanding is that 'top' should be 'top = bg.h/2 - font.h/2'.
2. I have tried other fonts and encountered a similar issue where the text is slightly shifted upwards when vertically centered.
3. Because text cannot be pre-sized at the pixel level before measurement, many times I center it around some special effects. When there's an error in vertical centering, this results in deviations in many of my effects. Of course, I've also tried using MeasureAdvance. Just use a fixed-size rectangular box to enclose the text, and you'll find that no matter how you center the rectangle, you can't successfully select the text perfectly.
```
    [Theory]
    [InlineData(300)]
    [InlineData(600)]
    [InlineData(1200)]
    public void TestBounds(int size)
    {
        const string text = "AA";
        var font = SystemFonts.CreateFont("Arial", Convert.ToInt32(size / 6.3));
        var bg = new Image<Rgba32>(size, size);
        var op = new RichTextOptions(font)
        {
            HorizontalAlignment = HorizontalAlignment.Center,
            TextAlignment = TextAlignment.Center,
            VerticalAlignment = VerticalAlignment.Center,
            Origin = new Vector2(size * 0.5f, size * 0.5f)
        };

        var ce = TextMeasurer.MeasureBounds(text, op);

        //Assuming the measured width and height are correct
        var topShould = bg.Height * 0.5f - ce.Height * 0.5f;
        var leftShould = bg.Width * 0.5f - ce.Width * 0.5f;
        var rightShould = bg.Width * 0.5f + ce.Width * 0.5f;
        var bottomShould = bg.Height * 0.5f + ce.Height * 0.5f;

        var top = ce.Top;
        var left = ce.Left;
        var right = ce.Right;
        var bottom = ce.Bottom;
        
        right.ShouldBeInRange(rightShould - 1, rightShould + 1, $"should {rightShould} => fact {right} diff {right - rightShould}");
        left.ShouldBeInRange(leftShould - 1, leftShould + 1, $"should {leftShould} => fact {left} diff {left - leftShould}");
        top.ShouldBeInRange(topShould - 1, topShould + 1, $"should {topShould} => fact {top} diff {top - topShould} ");
        bottom.ShouldBeInRange(bottomShould - 1, bottomShould + 1, $"should {bottomShould} => fact {bottom} diff {bottom - bottomShould}");
    }
```

> The larger the font, the greater the error.
![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/9151851/3304e727-c880-49ae-86c2-a5bc23d90f08)

![image](https://github.com/SixLabors/ImageSharp.Drawing/assets/9151851/c36c8404-cba5-4f98-8ce3-c11cfe9e6a4d)

![bdec2dfaf1534e7c9d8035b](https://github.com/SixLabors/ImageSharp.Drawing/assets/9151851/d0110158-28e9-4eb4-96a0-95206ee37e70)

![815567c4f82f4f28bd9ac37](https://github.com/SixLabors/ImageSharp.Drawing/assets/9151851/9cd74843-a46c-4e27-b696-076f89695e27)

**Comment by JimBobSquarePants** [comment, marked as answer]
Alignment is always measured based upon the line-height and advance (_which can collapse over each other for multiline text_). This means that the line is centered vertically not the glyph itself since glyphs are aligned within the advance box.

The following code demonstrates this well. Here I am drawing an `X` using vertical alignment with two red lines as a guide.

```c#
// Set variable dimensions
int w = 301;
int h = 301;
float hw = w / 2f;
float hh = h / 2f;

using Image<Rgba32> img = new(w, h, Color.White);

const string t = "X";
Font font = SystemFonts.CreateFont("Arial", h / 4);
RichTextOptions options = new(font)
{
	Origin = new PointF(hw, hh),
	VerticalAlignment = VerticalAlignment.Center,
	HorizontalAlignment = HorizontalAlignment.Center,
};

SolidBrush brush = new(Color.Red);
img.Mutate(x => x.DrawLine(brush, 1, new(hw, 0), new(hw, h)));
img.Mutate(x => x.DrawLine(brush, 1, new(0, hh), new(w, hh)));
img.Mutate(x => x.DrawText(options, t, Color.Black));

img.Save(@"C:\Users\james\Downloads\303.png");

```

If you do a side-by-side comparison (_CSS left, ImageSharp right_) with how a browser generates the same alignment, you can see that we match perfectly.

<img  alt="image" src="https://github.com/SixLabors/ImageSharp.Drawing/assets/385879/b930d418-7a3a-48a6-9ec7-9eb348db62a5">

The CSS can be found here.

https://jsfiddle.net/9qkgef17/

**Reply by inernoro** [reply]
You are professional, and I have learned new knowledge from you. Thank you

**Reply by JimBobSquarePants** [reply]
My pleasure
